File: 11.sorting/merge_sort.go
package main

func MergeSortHelper(input []string) []string {
	// Base case - nothing else to divide
	if len(input) < 2 {
		return input
	}

	// Divide array in halves
	middle := len(input) / 2

	leftRes := MergeSortHelper(input[:middle])
	rightRes := MergeSortHelper(input[middle:])

	left, right := 0, 0
	helperArr := []string{}

	for left < len(leftRes) && right < len(rightRes) {
		// Note: in case of a sort twice,
		// E.g. Sorting a table by any criteria
		// AND THEN sorting AGAIN by name,
		// "<=" will allow to preserve the previous criteria,
		// While also keeping the order by name.
		if leftRes[left] <= rightRes[right] {
			helperArr = append(helperArr, leftRes[left])
			left += 1
		} else {
			helperArr = append(helperArr, rightRes[right])
			right += 1
		}
	}

	helperArr = append(helperArr, leftRes[left:]...)
	helperArr = append(helperArr, rightRes[right:]...)

	return helperArr
}

func MergeSort(input []string) []string {
	input = MergeSortHelper(input)

	return input
}

// func main() {
// 	input := []string{"b", "c", "d", "a", "a", "e", "f", "z", "a", "y", "w", "h", "g", "y"}

// 	sorted := MergeSort(input)

// 	fmt.Printf("Merge Sort Result: %v\n", sorted)
// }

File: 11.sorting/merge_sort.py
class MergeSort:
    def __merge(self, left_input: list[str], right_input: list[str]) -> list[str]:
        pointer_left, pointer_right = 0, 0
        helper_arr = []

        # If you do "or" instead of the "and",
        # don't forget to do validations for "None"
        while pointer_left < len(left_input) and pointer_right < len(right_input):
            if left_input[pointer_left] < right_input[pointer_right]:
                helper_arr.append(left_input[pointer_left])
                pointer_left += 1
            else:
                helper_arr.append(right_input[pointer_right])
                pointer_right += 1

        helper_arr.extend(left_input[pointer_left:])
        helper_arr.extend(right_input[pointer_right:])

        return helper_arr

    def sort(self, input: list[str]) -> list[str]:
        # Base case - nothing else to divide
        if len(input) < 2:
            return input

        # Divide array in halves
        middle_index = len(input) // 2

        # Keep dividing in halves
        # Then do sort for both halves
        return self.__merge(self.sort(input[:middle_index]), self.sort(input[middle_index:]))


m = MergeSort()

input = ["b", "c", "d", "a", "a", "e", "f", "z", "a", "y", "w", "h", "g", "y"]
print(f"Merge Sort Result: {m.sort(input)}")

input = [5, 8, 3, 9, 4, 1, 7]
print(f"Merge Sort Result: {m.sort(input)}")
